pkg/services/ngalert/state/state.go

```diff
@@ -400,19 +395,38 @@ func resultKeepLast(state *State, rule *models.AlertRule, result eval.Result, lo
 	}
 }
 
-func (a *State) NeedsSending(resendDelay time.Duration) bool {
-	switch a.State {
-	case eval.Pending:
-		// We do not send notifications for pending states
+// NeedsSending returns true if the given state needs to be sent to the Alertmanager.
+// Reasons for sending include:
+// - The state has been resolved since the last notification.
+// - The state is firing and the last notification was sent at least resendDelay ago.
+// - The state was resolved within the resolvedRetention period, and the last notification was sent at least resendDelay ago.
+func (a *State) NeedsSending(resendDelay time.Duration, resolvedRetention time.Duration) bool {
+	if a.State == eval.Pending {
+		// We do not send notifications for pending states.
+		return false
+	}
+
+	// We should send a notification if the state has been resolved since the last notification.
+	if a.ResolvedAt.After(a.LastSentAt) {
```

Comment:
ok, this is supposed to cover the case when resolve notification was never sent and the interval expired.
There are two problems.
1. LastSentAt is not saved to db right now.  PR https://github.com/grafana/grafana/pull/89135 addresses it but does not address the situation fully because LastSentAt is updated _after_ instances are saved into database. (this is the most annoying part of the state management, IMO)
2. It is not limited. I.e. if instance starts in a couple of days, this will cause the alert to be created. So, effectively, we do not honor resolved retention.

Reply:
1. Agreed, the fact that we update LastSentAt after saving the state is awkward. If we're going to address this (and I think we should) it should be done in the [persistence PR](https://github.com/grafana/grafana/pull/89135)
2. Just to be clear, this should never create a `Firing` alert as only `Normal` states can have a non-zero `ResolvedAt`. That being said, it will send an expired alert to alertmanager, as intended with this change, as the associated `Firing` alert in the AM was never resolved. The only optimization here could be to check if the previous `Firing` alert has already expired, but we don't store the necessary information to do this from a warmed cache.